File: src/main/kafka/consumerService.ts
import type { Kafka, Consumer as KConsumer, Admin } from 'kafkajs'
import { randomUUID } from 'crypto'
import type { ConsumedMessage, ConsumerOptions, FetchMessagesOptions } from '../../shared/types'

export type MsgCallback = (msg: ConsumedMessage) => void

export class ConsumerService {
  /* ================================================================
   *  实时消费者 —— 常驻 consumer + 会话控制
   *
   *  策略（与消息浏览相同架构）：
   *   - 按 topic 维护常驻 consumer，consumer.run() 只启动一次
   *   - start：设置会话（onMsg 回调 + 分区过滤），必要时 seek
   *   - stop：清空会话，consumer 继续留在 group 中运行（消息丢弃）
   *   - 再次 start：无需 GROUP_JOIN，直接更新回调 + seek（~瞬间）
   *
   *  消息流转：
   *   eachMessage → 查 liveSessions（可能有多个同时活跃的 session）
   *               → 按 partition 过滤 → 调用对应 onMsg
   * ================================================================ */

  /** 按 topic 缓存的常驻 consumer */
  private liveConsumers = new Map<string, {
    consumer: KConsumer
    groupId: string
  }>()

  /** 活跃的消费会话（一个 topic 可以同时有多个 session） */
  private liveSessions = new Map<string, {    // key = consumerId
    topic: string
    partition?: number
    onMsg: MsgCallback
  }>()

  /** 辅助方法：从 KafkaMessage 提取 ConsumedMessage */
  private static toConsumedMessage(t: string, p: number, message: { offset: string; key?: Buffer | null; value?: Buffer | null; headers?: Record<string, Buffer | string | (Buffer | string)[] | undefined> | null; timestamp?: string | number }): ConsumedMessage {
    const headers: Record<string, string> = {}
    if (message.headers)
      for (const [k, v] of Object.entries(message.headers))
        if (v) headers[k] = Array.isArray(v) ? v.map(b => typeof b === 'string' ? b : b.toString('utf-8')).join(',') : typeof v === 'string' ? v : v.toString('utf-8')
    return {
      topic: t, partition: p, offset: message.offset,
      key: message.key?.toString('utf-8'),
      value: message.value?.toString('utf-8') ?? '',
      headers: Object.keys(headers).length > 0 ? headers : undefined,
      timestamp: Number(message.timestamp)
    }
  }

  /** 确保常驻 consumer 存在并运行 */
  private async ensureLiveConsumer(kafka: Kafka, topic: string, fromBeginning: boolean): Promise<{
    consumer: KConsumer
    isFirst: boolean
  }> {
    const cached = this.liveConsumers.get(topic)
    if (cached) {
      return { consumer: cached.consumer, isFirst: false }
    }

    const groupId = `live-${topic}`
    const consumer = kafka.consumer({ groupId })

    await consumer.connect()
    await consumer.subscribe({ topic, fromBeginning: true })
    this.liveConsumers.set(topic, { consumer, groupId })

    /* 启动常驻 fetch loop —— 只启动一次，永不 stop */
    consumer.run({
      autoCommit: false,
      eachMessage: async ({ topic: t, partition, message }) => {
        const msg = ConsumerService.toConsumedMessage(t, partition, message)
        /* 遍历所有活跃会话，投递消息 */
        this.liveSessions.forEach((session) => {
          if (session.topic !== t) return
          if (session.partition !== undefined && partition !== session.partition) return
          session.onMsg(msg)
        })
      }
    }).catch(() => { /* disconnect 时 reject，忽略 */ })

    console.log(`[live] 新建常驻 consumer topic=${topic}`)
    return { consumer, isFirst: true }
  }

  async start(kafka: Kafka, opts: ConsumerOptions, onMsg: MsgCallback): Promise<string> {
    const { topic } = opts
    const consumerId = randomUUID()
    const t0 = Date.now()

    const { consumer, isFirst } = await this.ensureLiveConsumer(kafka, topic, opts.fromBeginning ?? true)

    /* 注册会话 */
    this.liveSessions.set(consumerId, {
      topic,
      partition: opts.partition,
      onMsg
    })

    /* 首次创建需要等 GROUP_JOIN 再 seek；复用模式直接 seek */
    const seekTarget = opts.fromOffset ?? (opts.fromBeginning ? '0' : undefined)

    if (opts.partition !== undefined && seekTarget !== undefined) {
      if (isFirst) {
        await new Promise<void>((resolve) => {
          consumer.on(consumer.events.GROUP_JOIN, async () => {
            try { await consumer.seek({ topic, partition: opts.partition!, offset: seekTarget }) } catch { /**/ }
            resolve()
          })
        })
      } else {
        try { await consumer.seek({ topic, partition: opts.partition, offset: seekTarget }) } catch { /**/ }
      }
    }

    console.log(`[live] started id=${consumerId.slice(0, 8)} topic=${topic} isNew=${isFirst} t+${Date.now() - t0}ms`)
    return consumerId
  }

  async stop(consumerId: string): Promise<void> {
    const existed = this.liveSessions.delete(consumerId)
    if (existed) {
      console.log(`[live] stopped id=${consumerId.slice(0, 8)} (session removed, consumer stays)`)
    }
  }

  async stopAll(): Promise<void> {
    /* 清空所有会话 */
    this.liveSessions.clear()
    /* 断开所有常驻 live consumer */
    const liveEntries = Array.from(this.liveConsumers.values())
    for (const entry of liveEntries) {
      try { await entry.consumer.disconnect() } catch { /**/ }
    }
    this.liveConsumers.clear()
    /* 清理 fetch consumer */
    const fetchEntries = Array.from(this.fetchState.values())
    for (const f of fetchEntries) {
      try { await f.consumer.disconnect() } catch { /**/ }
    }
    this.fetchState.clear()
  }

  /* ================================================================
   *  消息浏览：常驻 consumer + seek，避免反复 GROUP_JOIN
   *
   *  策略：
   *   - 每个连接维护一个专用 fetch consumer，consumer.run() 只启动一次
   *   - 每次查询：seek 到目标 offset → 等待消息收齐 → 标记 done
   *   - consumer 始终留在 group中，不 stop / 不 disconnect
   *   - 首次 GROUP_JOIN ~3s，后续查询只需 seek + fetch（~50-200ms）
   * ================================================================ */

  private fetchState = new Map<string, {
    consumer: KConsumer
    topic: string
    running: boolean
    session: number      /* 递增会话 ID，用于标识当前查询 */
    collected: ConsumedMessage[]
    resolve?: () => void /* 当前等待的 Promise resolve */
    max: number
    partition?: number
  }>()

  /** 获取或创建 fetch 专用的常驻 consumer */
  private async ensureFetchConsumer(connId: string, kafka: Kafka, topic: string, partition?: number): Promise<{
    state: NonNullable<ConsumerService['fetchState'] extends Map<string, infer V> ? V : never>
    isFirst: boolean
  }> {
    const existing = this.fetchState.get(connId)
    if (existing && existing.topic === topic) {
      return { state: existing, isFirst: false }
    }

    /* 旧 consumer topic 不同或不存在，需要清理/创建 */
    if (existing) {
      try { await existing.consumer.disconnect() } catch { /**/ }
    }

    const consumer = kafka.consumer({
      groupId: `fetch-${connId.slice(0, 8)}`,
      maxWaitTimeInMs: 50,
    })

    const state: NonNullable<ConsumerService['fetchState'] extends Map<string, infer V> ? V : never> = {
      consumer,
      topic,
      running: false,
      session: 0,
      collected: [],
      max: 0,
      partition
    }

    await consumer.connect()
    await consumer.subscribe({ topic, fromBeginning: true })

    /* 启动常驻 fetch loop —— 只启动一次 */
    state.running = true
    consumer.run({
      autoCommit: false,
      eachMessage: async ({ topic: t, partition: p, message }) => {
        const s = this.fetchState.get(connId)
        if (!s || s.session === 0) return            /* 无活跃会话，丢弃 */
        if (s.partition !== undefined && p !== s.partition) return

        const headers: Record<string, string> = {}
        if (message.headers)
          for (const [k, v] of Object.entries(message.headers))
            if (v) headers[k] = typeof v === 'string' ? v : v.toString('utf-8')

        s.collected.push({
          topic: t, partition: p, offset: message.offset,
          key: message.key?.toString('utf-8'),
          value: message.value?.toString('utf-8') ?? '',
          headers: Object.keys(headers).length > 0 ? headers : undefined,
          timestamp: Number(message.timestamp)
        })

        if (s.collected.length >= s.max) {
          const resolve = s.resolve
          s.session = 0        /* 标记会话结束，后续消息直接丢弃 */
          s.resolve = undefined
          resolve?.()           /* 唤醒等待的 fetchMessages */
        }
      }
    }).catch(() => { /* disconnect 时 reject，忽略 */ })

    this.fetchState.set(connId, state)
    return { state, isFirst: true }
  }

  async fetchMessages(kafka: Kafka, opts: FetchMessagesOptions, poolAdmin?: Admin, connId?: string): Promise<ConsumedMessage[]> {
    const { topic, partition, offset, fromBeginning = false, limit = 50 } = opts
    const t0 = Date.now()

    /* ---- 1. admin 水位 ---- */
    let oo: Array<{ partition: number; high: string; low: string }>
    const needCleanup = !poolAdmin
    const admin = poolAdmin ?? kafka.admin()
    if (!poolAdmin) await admin.connect()
    try {
      oo = (await admin.fetchTopicOffsets(topic)).map((o) => ({
        partition: Number(o.partition), high: o.high, low: o.low
      }))
    } finally { if (needCleanup) await admin.disconnect().catch(() => {}) }

    /* ---- 2. 计算 seek 目标 ---- */
    const seeks: Array<{ partition: number; offset: string }> = []
    let expected = 0
    for (const o of oo) {
      if (partition !== undefined && o.partition !== partition) continue
      const hi = BigInt(o.high), lo = BigInt(o.low)
      let to: bigint
      if (offset !== undefined)      to = BigInt(offset)
      else if (!fromBeginning)       to = hi - BigInt(limit)
      else                           to = lo
      if (to < lo) to = lo
      const n = hi - to
      if (n > BigInt(0)) { expected += Number(n); seeks.push({ partition: o.partition, offset: to.toString() }) }
    }
    const max = Math.min(limit, expected)
    console.log(`[fetch] plan expected=${expected} max=${max} t+${Date.now() - t0}ms`)
    if (max <= 0) return []

    /* ---- 3. 获取常驻 consumer ---- */
    const cid = connId ?? `tmp-${randomUUID()}`
    const { state, isFirst } = await this.ensureFetchConsumer(cid, kafka, topic, partition)

    /* ---- 4. 首次启动需要等 GROUP_JOIN，后续直接 seek ---- */
    if (isFirst) {
      await new Promise<void>((resolve) => {
        const evt = state.consumer.events.GROUP_JOIN
        const onJoin = async (): Promise<void> => {
          for (const s of seeks)
            try { await state.consumer.seek({ topic, partition: s.partition, offset: s.offset }) } catch { /**/ }
          resolve()
        }
        state.consumer.on(evt, onJoin)
      })
      console.log(`[fetch] GROUP_JOIN + seek 完成 t+${Date.now() - t0}ms`)
    } else {
      /* 复用：直接 seek */
      for (const s of seeks)
        try { await state.consumer.seek({ topic, partition: s.partition, offset: s.offset }) } catch { /**/ }
      console.log(`[fetch] seek 完成 (复用) t+${Date.now() - t0}ms`)
    }

    /* ---- 5. 开始新会话，等待消息收齐 ---- */
    state.collected = []
    state.max = max
    state.partition = partition
    const sessionId = ++state.session

    const result = await Promise.race([
      /* 等待 eachMessage 回调收齐后 resolve */
      new Promise<ConsumedMessage[]>((resolve) => {
        state.resolve = () => resolve([...state.collected])
      }),
      /* 超时保底 */
      new Promise<ConsumedMessage[]>((resolve) =>
        setTimeout(() => {
          if (state.session === sessionId) {
            state.session = 0
            state.resolve = undefined
          }
          resolve([...state.collected])
        }, 10000)
      )
    ])

    console.log(`[fetch] 完成 ← ${result.length}/${max} 条, 耗时 ${Date.now() - t0}ms`)

    /* 非复用模式（无 connId）用完断开 */
    if (!connId) {
      try { await state.consumer.disconnect() } catch { /**/ }
      this.fetchState.delete(cid)
    }

    return result
  }
}

export const consumerSvc = new ConsumerService()

File: src-tauri/src/kafka/consumer_service.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use rdkafka::consumer::{Consumer, DefaultConsumerContext, StreamConsumer};
use rdkafka::client::DefaultClientContext;
use rdkafka::message::{BorrowedMessage, Headers, Message};
use serde::{Deserialize, Serialize};
use tauri::AppHandle;
use tokio::sync::{mpsc, RwLock};
use tracing::{info, warn};
use uuid::Uuid;

use crate::kafka::connection_manager::build_config_for_record;
use crate::store::connection_store::ConnectionRecord;

/// 消费者启动选项
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConsumerOptions {
    pub topic: String,
    pub partition: Option<i32>,
    #[serde(rename = "fromBeginning")]
    pub from_beginning: Option<bool>,
    #[serde(rename = "fromOffset")]
    pub from_offset: Option<String>,
}

/// 消息浏览选项
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FetchMessagesOptions {
    pub topic: String,
    pub partition: Option<i32>,
    pub offset: Option<String>,
    #[serde(rename = "fromBeginning")]
    pub from_beginning: Option<bool>,
    pub limit: Option<u32>,
}

/// 已消费的消息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConsumedMessage {
    pub topic: String,
    pub partition: i32,
    pub offset: String,
    pub key: Option<String>,
    pub value: String,
    pub headers: Option<HashMap<String, String>>,
    pub timestamp: i64,
}

/// 消费者会话
struct Session {
    topic: String,
    partition: Option<i32>,
    sender: mpsc::UnboundedSender<ConsumedMessage>,
}

/// 常驻消费者
struct LiveConsumer {
    _consumer: Arc<StreamConsumer<DefaultConsumerContext>>,
    cancel_tx: mpsc::Sender<()>,
}

/// 缓存的 fetch consumer（复用避免重复 GROUP_JOIN）
struct FetchConsumerCache {
    consumer: Arc<StreamConsumer<DefaultConsumerContext>>,
    topic: String,
    conn_id: String,
}

/// 消费者服务
pub struct ConsumerService {
    live_consumers: HashMap<String, LiveConsumer>,
    sessions: Arc<RwLock<HashMap<String, Session>>>,
    /// fetch consumer 缓存，按连接复用
    fetch_consumers: HashMap<String, FetchConsumerCache>,
    app_handle: Option<AppHandle>,
}

impl ConsumerService {
    pub fn new() -> Self {
        Self {
            live_consumers: HashMap::new(),
            sessions: Arc::new(RwLock::new(HashMap::new())),
            fetch_consumers: HashMap::new(),
            app_handle: None,
        }
    }

    pub fn set_app_handle(&mut self, handle: AppHandle) {
        self.app_handle = Some(handle);
    }

    /// 启动实时消费
    pub async fn start(
        &mut self,
        conn: &ConnectionRecord,
        opts: &ConsumerOptions,
    ) -> Result<(String, mpsc::UnboundedReceiver<ConsumedMessage>), String> {
        let topic = opts.topic.clone();
        let consumer_id = Uuid::new_v4().to_string();

        if !self.live_consumers.contains_key(&topic) {
            let mut config = build_config_for_record(conn)?;
            config.set("group.id", &format!("live-{}", topic));
            config.set("enable.auto.commit", "false");
            config.set("auto.offset.reset", "earliest");

            let consumer: StreamConsumer<DefaultConsumerContext> = config
                .create_with_context(DefaultConsumerContext)
                .map_err(|e| format!("Failed to create consumer: {e}"))?;

            consumer
                .subscribe(&[topic.as_str()])
                .map_err(|e| format!("Failed to subscribe: {e}"))?;

            let consumer_arc = Arc::new(consumer);
            let consumer_clone = Arc::clone(&consumer_arc);
            let (cancel_tx, mut cancel_rx) = mpsc::channel::<()>(1);

            let sessions_clone = Arc::clone(&self.sessions);
            let t = topic.clone();

            tokio::spawn(async move {
                loop {
                    tokio::select! {
                        _ = cancel_rx.recv() => {
                            info!("Live consumer for topic {} stopped", t);
                            break;
                        }
                        result = consumer_clone.recv() => {
                            match result {
                                Ok(msg) => {
                                    Self::dispatch_message(&sessions_clone, &msg).await;
                                }
                                Err(e) => {
                                    warn!("Consumer recv error: {}", e);
                                }
                            }
                        }
                    }
                }
            });

            self.live_consumers.insert(
                topic.clone(),
                LiveConsumer {
                    _consumer: consumer_arc,
                    cancel_tx,
                },
            );
            info!("Created live consumer for topic {}", topic);
        }

        let (tx, rx) = mpsc::unbounded_channel();
        self.sessions.write().await.insert(
            consumer_id.clone(),
            Session {
                topic: topic.clone(),
                partition: opts.partition,
                sender: tx,
            },
        );

        info!("Started session {} for topic {}", &consumer_id[..8], topic);
        Ok((consumer_id, rx))
    }

    /// 停止会话
    pub async fn stop(&mut self, consumer_id: &str) {
        self.sessions.write().await.remove(consumer_id);
        info!("Stopped session {}", &consumer_id[..8]);
    }

    /// 消息浏览：复用 fetch consumer（避免重复 GROUP_JOIN）
    pub async fn fetch_messages(
        &mut self,
        conn: &ConnectionRecord,
        conn_id: &str,
        admin: &rdkafka::admin::AdminClient<DefaultClientContext>,
        opts: &FetchMessagesOptions,
    ) -> Result<Vec<ConsumedMessage>, String> {
        let topic = opts.topic.clone();
        let limit = opts.limit.unwrap_or(50) as usize;
        let from_beginning = opts.from_beginning.unwrap_or(false);

        // 1. 获取水位（使用池化 admin，无额外连接开销）
        let metadata = admin
            .inner()
            .fetch_metadata(Some(&topic), Duration::from_secs(10))
            .map_err(|e| format!("Metadata error: {e}"))?;

        let topic_meta = metadata
            .topics()
            .first()
            .ok_or_else(|| format!("Topic not found: {topic}"))?;

        let mut seeks: Vec<(i32, i64)> = Vec::new();
        let mut expected = 0i64;

        for p in topic_meta.partitions() {
            let pid = p.id();
            if let Some(part) = opts.partition {
                if pid != part { continue; }
            }
            let (low, high) = admin
                .inner()
                .fetch_watermarks(&topic, pid, Duration::from_secs(5))
                .map_err(|e| format!("Watermark error: {e}"))?;

            let target = if let Some(ref off) = opts.offset {
                off.parse::<i64>().unwrap_or(low)
            } else if !from_beginning {
                (high - limit as i64).max(low)
            } else {
                low
            };
            let n = high - target;
            if n > 0 {
                expected += n;
                seeks.push((pid, target));
            }
        }

        let max = limit.min(expected as usize);
        if max == 0 {
            return Ok(vec![]);
        }

        let t0 = std::time::Instant::now();

        // 2. 获取或创建缓存的 fetch consumer（复用避免 GROUP_JOIN）
        let consumer = if let Some(cached) = self.fetch_consumers.get(conn_id) {
            if cached.topic != topic {
                // topic 变了，需要重新 subscribe
                cached.consumer.unsubscribe();
                cached
                    .consumer
                    .subscribe(&[topic.as_str()])
                    .map_err(|e| format!("Subscribe failed: {e}"))?;
            }
            Arc::clone(&cached.consumer)
        } else {
            let mut config = build_config_for_record(conn)?;
            config
                .set("group.id", &format!("fetch-{}", conn_id))
                .set("enable.auto.commit", "false")
                .set("auto.offset.reset", "earliest");

            let c: StreamConsumer<DefaultConsumerContext> = config
                .create_with_context(DefaultConsumerContext)
                .map_err(|e| format!("Failed to create consumer: {e}"))?;

            c.subscribe(&[topic.as_str()])
                .map_err(|e| format!("Subscribe failed: {e}"))?;

            // 等待首次 GROUP_JOIN
            let join_t0 = std::time::Instant::now();
            loop {
                let assignment = c
                    .assignment()
                    .map_err(|e| format!("Assignment error: {e}"))?;
                if assignment.count() > 0 {
                    break;
                }
                if join_t0.elapsed() > Duration::from_secs(15) {
                    return Err("GROUP_JOIN timeout".to_string());
                }
                tokio::time::sleep(Duration::from_millis(200)).await;
            }
            tracing::info!(
                "[fetch] GROUP_JOIN took {}ms (cached for reuse)",
                join_t0.elapsed().as_millis()
            );

            let consumer = Arc::new(c);
            self.fetch_consumers.insert(
                conn_id.to_string(),
                FetchConsumerCache {
                    consumer: Arc::clone(&consumer),
                    topic: topic.clone(),
                    conn_id: conn_id.to_string(),
                },
            );
            consumer
        };

        // 3. Seek 到目标 offset（复用模式下直接 seek，无需等 GROUP_JOIN）
        for (pid, offset) in &seeks {
            consumer
                .seek(&topic, *pid, rdkafka::Offset::Offset(*offset), Duration::from_secs(5))
                .map_err(|e| format!("Seek failed: {e}"))?;
        }

        // 4. Poll 收集消息
        let mut collected = Vec::new();
        let deadline = tokio::time::Instant::now() + Duration::from_secs(10);

        while collected.len() < max && tokio::time::Instant::now() < deadline {
            match tokio::time::timeout(Duration::from_secs(2), consumer.recv()).await {
                Ok(Ok(msg)) => {
                    if let Some(part) = opts.partition {
                        if msg.partition() != part { continue; }
                    }
                    let hdrs = msg.headers().map(|h| {
                        let mut map = HashMap::new();
                        for i in 0..h.count() {
                            let header = h.get(i);
                            map.insert(
                                header.key.to_string(),
                                String::from_utf8_lossy(header.value.unwrap_or(&[])).to_string(),
                            );
                        }
                        map
                    });
                    collected.push(ConsumedMessage {
                        topic: msg.topic().to_string(),
                        partition: msg.partition(),
                        offset: msg.offset().to_string(),
                        key: msg.key().map(|k| String::from_utf8_lossy(k).to_string()),
                        value: String::from_utf8_lossy(msg.payload().unwrap_or(&[])).to_string(),
                        headers: hdrs,
                        timestamp: msg.timestamp().to_millis().unwrap_or(0),
                    });
                }
                Ok(Err(_)) => break,
                Err(_) => break,
            }
        }

        tracing::info!(
            "[fetch] {} msgs from '{}' in {}ms (cached={})",
            collected.len(),
            topic,
            t0.elapsed().as_millis(),
            self.fetch_consumers.contains_key(conn_id)
        );
        Ok(collected)
    }

    /// 停止所有会话和消费者
    pub async fn stop_all(&mut self) {
        for (topic, live) in self.live_consumers.drain() {
            let _ = live.cancel_tx.send(()).await;
            info!("Stopped live consumer for topic {}", topic);
        }
        self.sessions.write().await.clear();
        // 清理 fetch consumer 缓存
        for (conn_id, cached) in self.fetch_consumers.drain() {
            cached.consumer.unsubscribe();
            info!("Stopped fetch consumer for connection {}", conn_id);
        }
    }

    /// 分发消息到匹配的会话
    async fn dispatch_message(
        sessions: &Arc<RwLock<HashMap<String, Session>>>,
        msg: &BorrowedMessage<'_>,
    ) {
        let sessions = sessions.read().await;
        if sessions.is_empty() { return; }

        let topic = msg.topic();
        let partition = msg.partition();

        let consumed = ConsumedMessage {
            topic: topic.to_string(),
            partition,
            offset: msg.offset().to_string(),
            key: msg.key().map(|k| String::from_utf8_lossy(k).to_string()),
            value: String::from_utf8_lossy(msg.payload().unwrap_or(&[])).to_string(),
            headers: msg.headers().map(|h| {
                let mut map = HashMap::new();
                for i in 0..h.count() {
                    let header = h.get(i);
                    map.insert(header.key.to_string(), String::from_utf8_lossy(header.value.unwrap_or(&[])).to_string());
                }
                map
            }),
            timestamp: msg.timestamp().to_millis().unwrap_or(0),
        };

        for session in sessions.values() {
            if session.topic != topic { continue; }
            if let Some(p) = session.partition {
                if p != partition { continue; }
            }
            let _ = session.sender.send(consumed.clone());
        }
    }
}

impl Default for ConsumerService {
    fn default() -> Self {
        Self::new()
    }
}
